### PR TITLE
Feat(eos_cli_config_gen): Add support for `monitor server radius`

### LINKED ISSUE #4273: Feat(eos_cli_config_gen): Add support for "monitor server radius"
### Discussed in https://github.com/aristanetworks/avd/discussions/4270

<div type='discussions-op-text'>

<sup>Originally posted by **chrisgerdener** July 26, 2024</sup>
I am back with another RADIUS related question/idea.

For our fabric we are using the "monitor server radius" to probe our RADIUS servers, as described in this Arista TOI https://www.arista.com/en/support/toi/eos-4-29-2f/17031-monitoring-radius-servers-via-arista-switches

As fas as I can see this is not yet supported in AVD, therefore we are implementing it with direct configuration through eos_cli variables.

I would like to propose that this be added as a new feature to eos_cli_config_gen

The data model could look like this:

```
monitor_server:
  radius:
    service:
      dot1x:
        probe:
          interval: <int, 1-1000>
          threshold_failure: <int, 1-255>
          method:
            status_server: <bool> # set to true by default, set to false for access_request probe
            access_request:
              username: <str>
              password: <str> # Password as type 7
                            
```

What does everyone think?                  
</div>

### PR DESCRIPTION
**Pull request by laxmikantchintakindi**
## Change Summary

Add support for `monitor server radius` in eos_cli_config_gen.

```
monitor_server_radius:
  service_dot1x: <bool>
  probe:
    interval: <int, 1-1000>
    threshold_failure: <int, 1-255>
    method: <str; "status-server" | "access-request">
    access_request: # Only relevant when method is 'access-request'
      username: <str>
      password: <str> # Password as type 7
```
## Related Issue(s)

Fixes #4273

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
Add support for `monitor server radius` in eos_cli_config_gen.

```
monitor_server_radius:
  service_dot1x: <bool>
  probe:
    interval: <int, 1-1000>
    threshold_failure: <int, 1-255>
    method: <str; "status-server" | "access-request">
    access_request: # Only relevant when method is 'access-request'
      username: <str>
      password: <str> # Password as type 7
```
### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+              - "access-request"
+          access_request:
+            type: dict
+            description: This can only be set when `method` is `access_request`. 
+            keys:
+              username:
+                type: str
+              password:
+                type: str
+              password_type:
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: please make it required keys

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{#
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2:1]: add the template for documentation as well

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-intended-config.j2 [hunk 3]
```diff
@@ -86,6 +86,8 @@
 {% include 'eos/match-list-input.j2' %}
 {# mcs client #}
 {% include 'eos/mcs-client.j2' %}
+{# monitor server radius #}
+{% include 'eos/monitor-server-radius.j2' %}
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-intended-config.j2:90]: can you confirm that monitor-server-radius config order with platform trident config

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-intended-config.j2:90]: Platform trident command are not configured on vEOS.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 1 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: Done.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 (hunk 2 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2:1]: Added

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-intended-config.j2 (hunk 3 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-intended-config.j2:90]: Confirm on trident. This order is correct.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+   probe interval {{ monitor_server_radius.probe.interval }} seconds
+{%     endif %}
+{%     if monitor_server_radius.probe.threshold_failure is arista.avd.defined %}
+   probe threshold failure {{ monitor_server_radius.probe.threshold_failure }}
+{%     endif %}
+{%     if monitor_server_radius.probe.method is arista.avd.defined("status-server") %}
+   probe method status-server
+{%     elif monitor_server_radius.probe.method is arista.avd.defined("access-request") %}
+{%         set access_request = monitor_server_radius.probe.access_request %}
+   probe method access-request username {{ access_request.username }} password {{ access_request.password_type }} {{ access_request.password }}
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: what if password_type is not defined

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: Can you please check the coverage and try to improve

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/monitor-server-radius.md
Comment (alexeygorbunov) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/monitor-server-radius.md:1]: We also need to add reference to this new file under `roles/eos_cli_config_gen/docs/input-variables.md`, otherwise it doesn't show up on the role's "Input Variables" mkdocs page

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 (hunk 4 above)
Comment (alexeygorbunov) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2:23]: it raises an error
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: jinja2.exceptions.UndefinedError: 'ansible.parsing.yaml.objects.AnsibleMapping object' has no attribute 'password_type'
fatal: [monitor-server-radius-1 -> localhost]: FAILED! => {"changed": false, "msg": "Error during plugin execution: 'ansible.parsing.yaml.objects.AnsibleMapping object' has no attribute 'password_type'"}
```

EoS allows to set `username` and `password` without specifying type of the password. In this case it treats it as type "0" (cleartext). Therefore if `username` and `password` are set but `type` is not set - we should default to type `0` (cleartext).

Comment (alexeygorbunov) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]:
```suggestion
   probe method access-request username {{ access_request.username }} password {{ access_request.password_type | arista.avd.default("0") }} {{ access_request.password }}
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                required: true
+              password:
+                type: str
+                required: true
+              password_type:
+                type: str
+                convert_types:
+                  - int
+                valid_values: ["0", "7", "8a"]
+                default: "7"
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml:55]:
```suggestion
                default: "0"
```

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 (hunk 4 above)
Comment (alexeygorbunov) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: Current approach assumes that we have access to `monitor_server_radius.probe.access_request` if `monitor_server_radius.probe.method == "access-request"` although this is not enforces by the schema.

Therefore setting `method: access-request` without setting `access_request` key causes:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: jinja2.exceptions.UndefinedError: 'ansible.parsing.yaml.objects.AnsibleMapping object' has no attribute 'access_request'
```

I would suggest to only render "access-request"-related code if both conditions are explicitly met

```suggestion
{%     elif monitor_server_radius.probe.method is arista.avd.defined("access-request") %}
{%         if monitor_server_radius.probe.access_request is arista.avd.defined() %}
{%             set access_request = monitor_server_radius.probe.access_request %}
   probe method access-request username {{ access_request.username }} password {{ access_request.password_type }} {{ access_request.password }}
{%         endif %}
```

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitor-server-radius.j2 [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+| Setting | Value |
+| ------- | ----- |
+{%         if monitor_server_radius.probe.interval is arista.avd.defined %}
+| Probe interval | {{ monitor_server_radius.probe.interval }} |
+{%         endif %}
+{%         if monitor_server_radius.probe.threshold_failure is arista.avd.defined %}
+| Threshold failure | {{ monitor_server_radius.probe.threshold_failure }} |
+{%         endif %}
+{%         if monitor_server_radius.probe.method is arista.avd.defined %}
+| Probe method | {{ monitor_server_radius.probe.method }} |
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitor-server-radius.j2:27]: Would it make sense to expose username (not sure about password) in case we use `method: access-request`?

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 (hunk 4 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: In all previous implementations, we have default password_type as 7 as we are accepting only type 7 passwords. If user wants clear text password he needs to choose password type 7. Reference schemas `tacacs_servers`, `monitor_telemetry_influx`.

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: Fixed.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/monitor-server-radius.md
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/monitor-server-radius.md:1]: Done.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius-1.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+### Monitor Server Radius Device Configuration
+
+```eos
+!
+monitor server radius
+   service dot1x
+   probe interval 100 seconds
+   probe threshold failure 100
+   probe method access-request username arista password 7 141600021F102B
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius-1.md]: please make sure this is removed from documentation by default using hide_password

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+{%     if monitor_server_radius.probe.threshold_failure is arista.avd.defined %}
+   probe threshold failure {{ monitor_server_radius.probe.threshold_failure }}
+{%     endif %}
+{%     if monitor_server_radius.probe.method is arista.avd.defined("status-server") %}
+   probe method status-server
+{%     elif monitor_server_radius.probe.method is arista.avd.defined("access-request") %}
+{%         if monitor_server_radius.probe.access_request.username is arista.avd.defined and
+              monitor_server_radius.probe.access_request.password is arista.avd.defined %}
+{%             set access_request = monitor_server_radius.probe.access_request %}
+   probe method access-request username {{ access_request.username }} password {{ access_request.password_type | arista.avd.default("7") }} {{ access_request.password }}
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]:
```suggestion
   probe method access-request username {{ access_request.username }} password {{ access_request.password_type | arista.avd.default("7") }} {{ access_request.password  | arista.avd.hide_passwords(hide_passwords }}
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius.md
Comment (alexeygorbunov) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius.md:1]: Is this file needed? There is no inventory host with such name

Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius.md:1]: nope it should be removed

**Review by gmuloc — DISMISSED** [review]

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2 (hunk 8 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-server-radius.j2]: Added.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius-1.md (hunk 7 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius-1.md]: Done

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius.md
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-server-radius.md:1]: Removed

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+          access_request:
+            type: dict
+            description: This can only be set when `method` is `access_request`.
+            keys:
+              username:
+                type: str
+                required: true
+              password:
+                type: str
+                description: Type 7 password.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: It is not a type-7 password if you select something else in password_type. Please look in other data models for inspiration for a better description.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      service_dot1x:
+        type: bool
+        description: Monitor servers used for 802.1X authentication.
+      probe:
+        type: dict
+        description: Settings for probe sent to the server.
+        keys:
+          interval:
+            type: int
+            description: Server probe period in seconds.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]:
```suggestion
            description: Server probe interval in seconds.
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/input-variables.md [hunk 11]
```diff
@@ -647,6 +647,12 @@ roles/eos_cli_config_gen/docs/tables/management-api-gnmi.md
 roles/eos_cli_config_gen/docs/tables/monitor-connectivity.md
 --8<--
 
+### Monitor Server Radius
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/input-variables.md]:
```suggestion
### Monitor server Radius
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-server-radius-2.yml
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-server-radius-2.yml:1]: please refactor the tests to add to host-1 and host-2 instead.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 9 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: Removed the description.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 10 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: Done.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/input-variables.md (hunk 11 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/input-variables.md]: Done.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-server-radius-2.yml
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-server-radius-2.yml:1]: Done.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-device-documentation.j2 [hunk 12]
```diff
@@ -121,6 +121,8 @@
 {% include 'documentation/quality-of-service.j2' %}
 {# InfluxDB Telemetry #}
 {% include 'documentation/monitor-telemetry-influx.j2' %}
+{# Monitor Server Radius #}
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-device-documentation.j2]: There is a file called `monitoring.j2` which contains all the other similar monitor features. Please move it there and update the headings to being from ###.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            valid_values:
+              - "status-server"
+              - "access-request"
+          access_request:
+            type: dict
+            description: This can only be set when `method` is `access_request`.
+            keys:
+              username:
+                type: str
+                required: true
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]:
```suggestion
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+              password:
+                type: str
+                required: true
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]:
```suggestion
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+              - "access-request"
+          access_request:
+            type: dict
+            description: This can only be set when `method` is `access_request`.
+            keys:
+              username:
+                type: str
+                required: true
+              password:
+                type: str
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml:47]:
```suggestion
                type: str
                description: Encrypted password using the password_type.
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 15 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml:47]: Done.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 14 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: Removed.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 13 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml]: Removed

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-device-documentation.j2 (hunk 12 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos-device-documentation.j2]: Done.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitoring.j2 [hunk 16]
```diff
@@ -62,4 +62,6 @@
 {%     include 'documentation/trackers.j2' %}
 {## Monitor Telemetry Postcard #}
 {%     include 'documentation/monitor-telemetry-postcard-policy.j2' %}
+{## Monitor Server Radius #}
```
Comment (alexeygorbunov) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitoring.j2:66]: `if` statement on lines 18-43 needs to be adjusted to match if `monitor_server_radius` is defined. Right now `host2`'s documentation is missing section for `Monitor Server Radius Summary`

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 15 above)
Comment (alexeygorbunov) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml:47]: Should we remove "0" from the list of valid `password_types` options as this would not be a hashed value?

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitoring.j2 (hunk 16 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitoring.j2:66]: Fixed. Thanks

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitor-server-radius.j2 (hunk 6 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/documentation/monitor-server-radius.j2:27]: I have not seen username in any document but I agree that it would be better to add.

**Review by alexeygorbunov — APPROVED** [review]
LGTM

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml (hunk 15 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/monitor_server_radius.schema.yml:47]: no clear text password is ok as well.

**Review by ClausHolbechArista — APPROVED** [review]
